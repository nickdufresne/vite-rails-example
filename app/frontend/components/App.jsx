import React from 'react'

function App() {
  const [data, setData] = React.useState([]);
  React.useEffect(() => {
    fetch('/api/books')
    .then((res) => res.json())
    .then((data) => {setData(data)});
  }, []);
  return (
    <div>
      Books:
      {
        data.map(({title, author, description}) =>
          <div className="book">
            <div> Title: {title}</div>
            <div> Author: {author}</div>
            <div> Description: {description}</div>
          </div>
        )
      }
    </div>
  );
}

export default App;
