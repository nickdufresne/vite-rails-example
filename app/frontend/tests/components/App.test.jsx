import { render, act, waitFor } from '@testing-library/react';
import { describe, expect, it, vi, beforeEach } from 'vitest'
import App from '../../components/App';

const stubbedResponses = []

function stubFetchResponse(url, response) {
  stubbedResponses.push([url, response]);
}

async function fetch(resource, options) {
  const response = stubbedResponses.shift();
  if (resource == response[0]) {
    return { json: () => new Promise((resolve) => resolve(response[1])) }
  } else {
    throw new Error(`URL mismatch: ${response[0]} != ${resource}`);
  }
}

global.fetch = vi.fn(fetch);

describe("App", async() => {

  it("renders", async() => {
    const booksApiResponse = [
      {
        title: 'Book One',
        author: 'Author One',
        description: 'Description',
      },
    ]

    stubFetchResponse("/api/books", booksApiResponse);

    const { container } = render(
      <App />,
    );

    expect(container).toMatchSnapshot();

    await act(async () => {
      return Promise.resolve();
    });

    expect(container).toMatchSnapshot();
  });
});
