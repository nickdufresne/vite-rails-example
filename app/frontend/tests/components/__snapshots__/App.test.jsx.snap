// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders 1`] = `
<div>
  <div>
    Books:
  </div>
</div>
`;

exports[`App > renders 2`] = `
<div>
  <div>
    Books:
    <div
      class="book"
    >
      <div>
         Title: 
        Book One
      </div>
      <div>
         Author: 
        Author One
      </div>
      <div>
         Description: 
        Description
      </div>
    </div>
  </div>
</div>
`;
